api/v2/serializers/post/token_update.py

```diff
@@ -23,30 +24,61 @@ def validate(self, data):
+        # It should be here as it is part of the framework.
+        return data
 
-    def create(self, validated_data):
-        identity = self._find_identity_match(validated_data['provider'], validated_data['username'], validated_data['project_name'])
+    def create(self, data):
+        logger.info("TokenUpdateSerializer::create called")
+        identity = self._find_identity_match(data['provider'], data['username'], data.get('project_name'))
         if not identity:
-            identity = self._create_identity(validated_data['provider'], validated_data['username'], validated_data['project_name'], validated_data['token'])
-            return identity
+            identity = self._create_identity(data['provider'],
+                                             data['username'],
+                                             data.get('project_name'),
+                                             data.get('token'))
+        logger.info("cred -> ex_force_auth_token = %s" % (data.get('token')))
+        identity.update_credential(identity, 'key', data['username'], replace=True)
+        identity.update_credential(identity, 'ex_force_auth_token', data.get('token'), replace=True)
+        identity.update_credential(identity,
+                                   'ex_force_auth_url',
+                                   settings.AUTHENTICATION['KEYSTONE_SERVER'],
+                                   replace=True)
 
-        token_cred = identity.credential_set.filter(key='ex_force_auth_token').first()
-        if token_cred:
-            token_cred.value = validated_data['token']
-            token_cred.save()
-        else:
-            identity.credential_set.create(key='ex_force_auth_token', value=validated_data['token'])
+        auth_url = settings.AUTHENTICATION['KEYSTONE_SERVER'] + '/v3'
+        response = requests.get(auth_url + '/auth/tokens',
+                                headers={'x-auth-token': data['token'], 'x-subject-token': data['token']})
+        try:
+            catalog = json.loads(response.text)
+            project = catalog['token']['project']
+        except KeyError:
+            raise serializers.ValidationError("Invalid token passed")
+        endpoint_catalog = catalog['token']['catalog']
+        compute = None
+        for listing in endpoint_catalog:
+            if listing['type'] == 'compute':
+                compute = listing
+        if not compute:
+            raise serializers.ValidationError("Cannot find compute endpoint catalog")
+        # need to go through the endpoints to find the public one
+        compute_url = None
+        for ep in compute['endpoints']:
+            if ep['interface'] == 'public':
+                compute_url = ep['url']
+        if not compute_url:
+            raise serializers.ValidationError("Cannot find a public compute endpoint url")
+        identity.update_credential(identity, 'ex_force_base_url', compute_url, replace=True)
```

Comment:
The base_url is specified in the catalog as passed back from keystone when anyone logs it.  This makes the configuration simpler as there is one fewer urls to.  We will still need to process the keystone catalog to get this url.

As libcloud has different behaviors based on if ex_force_auth_url/ex_force_base_url exists, I chose to put both of these in the user credentials.  This way, if there are any accounts that are not using the ex_force_auth_url/ex_force_base_url for that provider, we don't need to have special logic to distinguish the difference between them.